src/Raven.Client/Http/RequestExecuter.cs

```diff
@@ -296,41 +317,91 @@ private void UpdateTopologyCallback(object _)
         private async Task HandleServerDown<TResult>(ServerNode node, JsonOperationContext context, RavenCommand<TResult> command,
             HttpRequestException e)
         {
-
             if (command.FailedNodes == null)
                 command.FailedNodes = new HashSet<ServerNode>();
 
+            node.IsFailed = true;
             command.FailedNodes.Add(node);
 
-            var failoverNode = GetFailoverNode(command);
-
-            if (failoverNode != null)
-            {
-                await ExecuteAsync(failoverNode, context, command);
-                return;
-            }
-
-            throw new HttpRequestException("Tried all nodes in the cluster but failed getting a response", e);
+            var failoverNode = ChooseNodeForRequset(command, e);
+            await ExecuteAsync(failoverNode, context, command);
         }
 
-        private ServerNode GetFailoverNode<T>(RavenCommand<T> command)
+        private ServerNode ChooseNodeForRequset<T>(RavenCommand<T> command, HttpRequestException exception = null)
         {
-            //TODO: implement failover policies
             var topology = _topology;
 
-            if (command.FailedNodes == null)
-                command.FailedNodes = new HashSet<ServerNode>();
-
             var leaderNode = topology.LeaderNode;
-            if (command.FailedNodes.Contains(leaderNode) == false)
-                return leaderNode;
 
-            foreach (var node in topology.Nodes)
+            if (command.IsReadRequest)
+            {
+                if (topology.ReadBehavior == ReadBehavior.Leader)
+                {
+                    if (command.IsFailedWithNode(leaderNode) == false)
+                        return leaderNode;
+                    throw new HttpRequestException("Leader not was failed to make this request. The current ReadBehavior is set to Leader to we won't failover to a differnt node.", exception);
+                }
+
+                if (topology.ReadBehavior == ReadBehavior.All)
+                {
+                    if (leaderNode.IsFailed == false && command.IsFailedWithNode(leaderNode) == false)
+                        return leaderNode;
+
+                    // TODO: Should we choose nodes here by rate value as for SLA?
+                    foreach (var node in topology.Nodes)
+                    {
+                        if (node.IsFailed == false && command.IsFailedWithNode(leaderNode) == false)
```

Comment:
I think that you want to pass node to command.IsFailedWithNode not the leaderNode ?